### PR TITLE
fix: introduce postgres advisory locks to avoid data concurrency issues

### PR DESCRIPTION
**Pull request by alexis-kochetov**

### CODE REVIEW

File: src/dotnet/Chat.Service/ChatsBackend.cs [hunk 1]
```diff
@@ -1515,6 +1524,7 @@ public virtual async Task OnUpdateReadPositionsStat(
             }
         }
         else {
+            // await dbContext.ReadPositionsStats.Lock(chatId, cancellationToken).ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/Chat.Service/ChatsBackend.cs]: remove

File: src/dotnet/Db/DbSetExt.cs [hunk 2]
```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
...
+        var keyHash = (long)(key?.GetHashCode() ?? 239);
+        var lockKey = (keyHash << 32) ^ HashCode.Combine(typeof(TEntity).GetHashCode());
+        var timeout = context.Database.GetCommandTimeout();
+        // TODO(AK): find a way to resolve DBSettings instance
+        context.Database.SetCommandTimeout(30);
+        try {
+            await context.Database.ExecuteSqlAsync($"SELECT pg_advisory_xact_lock({lockKey});", cancellationToken);
+        }
+        catch (Exception ex) {
+            // TODO(AK): find a way to resolve a logger
```
Comment (crui3er) [inline src/dotnet/Db/DbSetExt.cs]: Use StaticLog

File: src/dotnet/Db/DbSetExt.cs [hunk 3]
```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
...
+            await Console.Error.WriteLineAsync("Lock failed: " + ex.Message);
```
Comment (crui3er) [inline src/dotnet/Db/DbSetExt.cs]: Add details to log: DbSet, key

File: src/dotnet/Db/DbSetExt.cs [hunk 4]
```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
             throw new ArgumentOutOfRangeException(nameof(key), "Key must be greater than zero.");
         return set.FindAsync(DbKey.Compose(key), cancellationToken);
     }
+
+    public static async Task Lock<TEntity, TKey>(
+        this DbSet<TEntity> set,
+        TKey key,
+        CancellationToken cancellationToken)
```
Comment (crui3er) [inline src/dotnet/Db/DbSetExt.cs:34]: Looks like cancellationToken is not used. Do we need it?

File: src/dotnet/Db/DbSetExt.cs [hunk 5]
```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
...
+        TKey key,
+        TArg0 arg0,
+        CancellationToken cancellationToken)
+        where TEntity : class
+    {
+        var context = set.GetDbContext();
+        var keyHash = (long)(key?.GetHashCode() ?? 239);
+        var lockKey = (keyHash << 32) ^ HashCode.Combine(typeof(TEntity).GetHashCode(), arg0);
+        var timeout = context.Database.GetCommandTimeout();
+        try {
```
Comment (crui3er) [inline src/dotnet/Db/DbSetExt.cs]: It seems all Lock methods constists of 2 parts: lockKey calculation and  executing lock command. Extract executing lock command to a separate method and reuse it.

File: src/dotnet/Db/DbSetExt.cs [hunk 6]
```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
...
+        this DbSet<TEntity> set,
+        TKey key,
+        CancellationToken cancellationToken)
+        where TEntity : class
+    {
+        var context = set.GetDbContext();
+        var keyHash = (long)(key?.GetHashCode() ?? 239);
+        var lockKey = (keyHash << 32) ^ HashCode.Combine(typeof(TEntity).GetHashCode());
+        var timeout = context.Database.GetCommandTimeout();
+        // TODO(AK): find a way to resolve DBSettings instance
```
Comment (crui3er) [inline src/dotnet/Db/DbSetExt.cs]: You can make DBSettings.LockTimeout property static so far or move this to constants.

**Review by crui3er — CHANGES_REQUESTED** [review]

File: src/dotnet/Db/DbSetExt.cs (hunk 4 above)
Comment (alexis-kochetov) [reply src/dotnet/Db/DbSetExt.cs:34]: image

**Pull request merged**